### PR TITLE
feat(NcSelect): Allow to filter users by email notation

### LINKED ISSUE #4491: NcSelect does not recognize email addresses when angle bracket notation is used
When searching a user with an email address while the email is entered in angle brackets, NcSelect should not filter out the matching options.

John john@example.org entered ✔️  | John <john@example.org> entered ❌ 
-|-
grafik|grafik

Expected behavior: When entering `John <john@example.org>` the user should be found as well as when entering without brackets. 

I would call it a bug, since upon a user's report, this worked in earlier versions, but I can't confirm this. https://github.com/nextcloud/polls/issues/3036

### PR DESCRIPTION
**Pull request by susnux**
### ☑️ Resolves

* Fix #4491 

Allows to filter by the email notation like `Jane <j.doe@example.com>`

### 🏁 Checklist

- [x] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable

### CODE REVIEW

**Comment by raimund-schluessler** [comment]
Might be, that I do something wrong, but it doesn't seem to work for me in the styleguide:
grafik

**Comment by susnux** [comment]
> Might be, that I do something wrong, but it doesn't seem to work for me in the styleguide

@raimund-schluessler Ah yes there was an issue if there is no `subname` defined, fixed that:

https://github.com/nextcloud-libraries/nextcloud-vue/assets/1855448/a6da226c-8f2f-45ce-861a-c4d09c37ebdc

File: src/components/NcSelect/NcSelect.vue [hunk 1]
```diff
@@ -951,14 +951,19 @@ export default {
...
 			if (this.filterBy !== null) {
 				return this.filterBy
 			}
 			if (this.userSelect) {
 				return (option, label, search) => {
-					return (`${label} ${option.subname}` || '')
-						.toLocaleLowerCase()
-						.indexOf(search.toLocaleLowerCase()) > -1
+					const match = search.match(EMAIL_NOTATION)
+					return (match && option.subname && option.subname.toLocaleLowerCase().indexOf(match[1].toLocaleLowerCase()) > -1)
```
Comment (raimund-schluessler) [inline src/components/NcSelect/NcSelect.vue:963]:
```suggestion
					return (match && option.subname?.toLocaleLowerCase?.().indexOf(match[1].toLocaleLowerCase()) > -1)
```

**Review by raimund-schluessler — APPROVED** [review]
Works now as expected (in the docs).

**Review by Pytal — COMMENTED** [review]
Is it possible to have the matching chars bolded like when typing an email without angle brackets?

**Comment by susnux** [comment]
> Is it possible to have the matching chars bolded like when typing an email without angle brackets?

Yes but need to adjust `NcListItemIcon` then.

**Comment by susnux** [comment]
@Pytal done


https://github.com/nextcloud-libraries/nextcloud-vue/assets/1855448/9272ff6e-4107-4a5e-b562-c6cdb0c68e4d

File: src/components/NcSelect/NcSelect.vue (hunk 1 above)
Comment (susnux) [reply src/components/NcSelect/NcSelect.vue]: Done

**Comment by raimund-schluessler** [comment]
> > So, when typing <olivia@example in NcSelect, do the matching in NcSelect and set this.search to olivia@example without the token. This way, the highlighting should work, and we wouldn't need to duplicate the matching code in NcListItemIcon.
> 
> 
> 
> Yes but when typing `Olivia <olivia@example` **Olivia** would not be highlighted.
> 
> So I think this is also a feature on the `NcListItemIcon`

Ah, yes, I see. Valid point.

File: src/components/NcListItemIcon/NcListItemIcon.vue [hunk 2]
```diff
@@ -314,6 +314,18 @@ export default {
...
+		/**
+		 * Seperates the search property into two parts, the first one is the search part on the name, the second on the subname.
+		 * @return {[string, string]}
+		 */
+		searchParts() {
+			// Match the email notation like "Jane <j.doe@example.com>" with the email address as matching group
+			const EMAIL_NOTATION = /^([^<]*)<([^>]+)>?$/
+
+			const match = this.search.match(EMAIL_NOTATION)
+			return this.isNoUser || !match ? [this.search, this.search] : [match[1].trim(), match[2]]
```
Comment (Pytal) [inline src/components/NcListItemIcon/NcListItemIcon.vue:327]:
```suggestion
			if (this.isNoUser) {
				return [this.search, this.search]
			}
			return !match ? [this.search, this.search] : [match[1].trim(), match[2]]
```

Comment (susnux) [reply src/components/NcListItemIcon/NcListItemIcon.vue]: done

**Review by Pytal — APPROVED** [review]

**Comment by dartcafe** [comment]
@Pytal Any chance to get this backported to v7?

**Comment by Pytal** [comment]
> @Pytal Any chance to get this backported to v7?

We don't normally backport features @dartcafe, so I'd recommend using the latest v8 of the lib :)

**Pull request merged**